jlm/llvm/opt/RvsdgTreePrinter.cpp

```diff
@@ -115,6 +118,78 @@ RvsdgTreePrinter::AnnotateNumRvsdgNodes(
   annotateRegion(*rvsdg.root());
 }
 
+void
+RvsdgTreePrinter::AnnotateNumMemoryStateInputsOutputs(
+    const rvsdg::graph & rvsdg,
+    util::AnnotationMap & annotationMap)
+{
+  static std::string_view argumentLabel("NumMemoryStateTypeArguments");
+  static std::string_view resultLabel("NumMemoryStateTypeResults");
+  static std::string_view inputLabel("NumMemoryStateTypeInputs");
+  static std::string_view outputLabel("NumMemoryStateTypeOutputs");
+
+  std::function<void(const rvsdg::region &)> annotateRegion = [&](const rvsdg::region & region)
+  {
+    size_t numMemoryStateArguments = 0;
+    for (size_t n = 0; n < region.narguments(); n++)
+    {
+      auto argument = region.argument(n);
+      if (rvsdg::is<MemoryStateType>(argument->type()))
+      {
+        numMemoryStateArguments++;
+      }
+    }
+    annotationMap.AddAnnotation(&region, { argumentLabel, numMemoryStateArguments });
+
+    size_t numMemoryStateResults = 0;
+    for (size_t n = 0; n < region.nresults(); n++)
+    {
+      auto result = region.result(n);
+      if (rvsdg::is<MemoryStateType>(result->type()))
+      {
+        numMemoryStateResults++;
+      }
+    }
+    annotationMap.AddAnnotation(&region, { resultLabel, numMemoryStateResults });
+
+    for (auto & node : region.nodes)
+    {
+      if (auto structuralNode = dynamic_cast<const rvsdg::structural_node *>(&node))
+      {
+        size_t numMemoryStateInputs = 0;
+        for (size_t n = 0; n < structuralNode->ninputs(); n++)
+        {
+          auto input = structuralNode->input(n);
+          if (rvsdg::is<MemoryStateType>(input->type()))
+          {
+            numMemoryStateInputs++;
+          }
+        }
+        annotationMap.AddAnnotation(structuralNode, { inputLabel, numMemoryStateInputs });
+
+        size_t numMemoryStateOutputs = 0;
+        for (size_t n = 0; n < structuralNode->noutputs(); n++)
+        {
+          auto output = structuralNode->output(n);
+          if (rvsdg::is<MemoryStateType>(output->type()))
+          {
+            numMemoryStateOutputs++;
+          }
+        }
+        annotationMap.AddAnnotation(structuralNode, { outputLabel, numMemoryStateOutputs });
+
+        for (size_t n = 0; n < structuralNode->nsubregions(); n++)
+        {
+          auto subregion = structuralNode->subregion(n);
+          annotateRegion(*subregion);
+        }
+      }
+    }
+  };
```

Comment:
There is a lot of repetition in here. However, in order to remove this, it would be necessary to add better support for iterators/iterator ranges to the base classes.